hadoop-tools/hadoop-aliyun/src/main/java/org/apache/hadoop/fs/aliyun/oss/AliyunOSSInputStream.java

```diff
@@ -316,4 +349,248 @@ public boolean seekToNewSource(long targetPos) throws IOException {
+        child.getData().complete(buffer);
+        prev = child;
+      }
+    }
+  }
+
+  /**
+   * Read data from OSS for this range and populate the buffer.
+   * @param range range of data to read.
+   * @param buffer buffer to fill.
+   */
+  private void readSingleRange(FileRange range, ByteBuffer buffer) {
+    LOG.debug("Start reading range {} from path {} ", range, key);
+    InputStream objectRange = null;
+    try {
+      long position = range.getOffset();
+      int length = range.getLength();
+      objectRange = store.retrieve(key, position, position + length - 1);
+      populateBuffer(range, buffer, objectRange);
+      range.getData().complete(buffer);
+    } catch (Exception ex) {
+      LOG.warn("Exception while reading a range {} from path {} ", range, key, ex);
+      range.getData().completeExceptionally(ex);
+    } finally {
+      IOUtils.cleanupWithLogger(LOG, objectRange);
+    }
+  }
+
+  /**
+   * Read the data from OSS for the bigger combined file range and update all the
+   * underlying ranges.
+   * @param combinedFileRange big combined file range.
+   * @param allocate method to create byte buffers to hold result data.
+   */
+  private void readCombinedRangeAndUpdateChildren(CombinedFileRange combinedFileRange,
+                                                  IntFunction<ByteBuffer> allocate) {
+    LOG.debug("Start reading combined range {} from path {} ", combinedFileRange, key);
+    InputStream objectRange = null;
+    try {
+      long position = combinedFileRange.getOffset();
+      int length = combinedFileRange.getLength();
+      objectRange = store.retrieve(key, position, position + length - 1);
+      populateChildBuffers(combinedFileRange, objectRange, allocate);
+    } catch (Exception ex) {
+      LOG.debug("Exception while reading a range {} from path {} ", combinedFileRange, key, ex);
+      for(FileRange child : combinedFileRange.getUnderlying()) {
+        child.getData().completeExceptionally(ex);
+      }
+    } finally {
+      IOUtils.cleanupWithLogger(LOG, objectRange);
+    }
+    LOG.debug("Finished reading range {} from path {} ", combinedFileRange, key);
+  }
+
+  /**
+   * {@inheritDoc}
+   * Vectored read implementation for {@link AliyunOSSInputStream}.
+   * @param ranges the byte ranges to read.
+   * @param allocate the function to allocate ByteBuffer.
+   * @throws IOException IOE if any.
+   */
+  @Override
+  public void readVectored(List<? extends FileRange> ranges,
+                           IntFunction<ByteBuffer> allocate) throws IOException {
+    LOG.debug("Starting vectored read on path {} for ranges {} ", key, ranges);
+    checkNotClosed();
+    if (stopVectoredIOOperations.getAndSet(false)) {
+      LOG.debug("Reinstating vectored read operation for path {} ", key);
+    }
+    List<? extends FileRange> sortedRanges = validateAndSortRanges(ranges,
+        Optional.of(contentLength));
+    for (FileRange range : ranges) {
+      validateRangeRequest(range);
```

Comment:
the validateAndSortRanges operation will now do this, so this is not needed

Reply:
OK